MovingBlocks/Terasology
# Input modifier

### Contains

This PR works towards solving #2551.

What was added:

* Ability to bind keys with modifiers(Alt, Shift, or Control) with @DefaultBinding annotation and ingame input binding changer
* Json deserialization for modified inputs
* A new extension of Input that allows for a modifier to be attached to it
* Alteration of some current existing input functions to allow for use of new modified input as well
* A bunch of logic to the InputSystem that accounts for these modifier keys.


Currently you cannot bind a key to just control, alt, or shift unless it is from a defaultBinding, but everything still works as expected if something is bound to those keys due to defaultBinding.

New variation on Default binding: 

`@DefaultBinding(type = InputType.KEY, id = Keyboard.KeyId.R)`

Can now also be done if a default binding with a modifier is wanted:

`@DefaultBinding(type = InputType.KEY, id = Keyboard.KeyId.R, mod = InputModified.Modifier.ALT)`


Currently designed that if multiple modifier keys are pressed, all variations are checked, but no multiple modifiers are pressed:
* If Alt, Shift, W, and E are being pressed then it is same as the keys (Alt), (Shift), (Alt + W), (Alt + E), (Shift + W), and (Shift + E) being pressed
* When binding with inputs if multiple modifiers are pressed they take priority of alt > control > shift

### How to test

This PR changes a bit of the logic behind the inputSystem and therefore is very expansive for testing, I would not be surprised if I missed a certain test or what not, but I will list below everything that I tested as I developed. 

* Having a default binding as any of the modifier keys or the modifier "none"
* Reset to the above binding
* Set a binding based on alt, control, or shift and test all possible holding combinations(solo, alt +control, alt + shift, all three, etc)
* Test for if there is overlap on setting an input that matches (alt + w vs alt + w)
* Test for overlap on inputs that match only on modifier (alt + w vs alt + r)
* Test for overlap on inputs that match only on key (alt + w vs w)
* Test for overlap on inputs that match with same key, different modifier (alt +w vs shift + w)
* Test all of the above in game on actual binding inputs
* Test for proper interaction between pressing and unpressing modifier or normal keys

   Having something bound to W and Alt + W. Pressing alt by self -> nothing, then w -> (Alt + W) binding, release Alt -> stop (Alt + W) and trigger (W). Hold just W -> (W), press alt -> Stop (W) and start (Alt + W), release W -> Stop (Alt + W), but don't start (W)
  
  
* Test inputs on different types of input (Inputs that only check press portion, and inputs that check repeats, etc)
* Good ol button spamming and seeing a "correct" response across output with no crashing


### Outstanding before merging

Currently only with Keys. Could be extended to controller/mouse inputs as well, but didn't really seem to fit for the current design. But, this does allow for a very good idea on how to implement say shift clicking items in an inventory and what not. 


Still new to programming in larger projects and so will probably have some slight coding oddities that would need to be double checked before fully merging, but just let me know and I will be glad to make fixes/work towards different methods as needed.
